airbyte-integrations/connectors/destination-keen/src/main/java/io/airbyte/integrations/destination/keen/KeenTimestampService.java

```diff
@@ -0,0 +1,187 @@
+        // so it won't be parsed for future messages.
+        LOGGER.info("Unable to parse timestamp field: " + cursorField.name);
+        streamCursorFields.remove(streamName);
+        injectTimestamp(data, Instant.ofEpochMilli(message.getRecord().getEmittedAt()).toString());
+      }
+    }
+    else {
+      injectTimestamp(data, Instant.ofEpochMilli(message.getRecord().getEmittedAt()).toString());
+    }
+    return data;
+  }
+
+  private void injectTimestamp(JsonNode data, String timestamp) {
+    ObjectNode root = ((ObjectNode) data);
+    root.set("keen", JsonNodeFactory.instance.objectNode().put("timestamp", timestamp));
+  }
+
+  private String parseTimestamp(CursorField cursorField, JsonNode data) {
+    return switch (cursorField.type) {
+      case NUMBER -> dateFromNumber(data.get(cursorField.name).asLong());
+      case STRING -> parser
+          .parse(data.get(cursorField.name).asText())
+          .get(0).getDates()
+          .get(0)
+          .toInstant()
+          .toString();
+      default -> throw new IllegalStateException("Unexpected value: " + cursorField.type);
+    };
+  }
+
+  private String dateFromNumber(Long timestamp) {
+    // if cursor value is below given threshold, assume that it's not epoch timestamp but ordered id
+    if (timestamp < SECONDS_FROM_EPOCH_THRESHOLD) {
+      throw new IllegalArgumentException("Number cursor field below threshold: " + timestamp);
+    }
+    // if cursor value is above given threshold, then assume that it's Unix timestamp in milliseconds
+    if (timestamp > MILLIS_FROM_EPOCH_THRESHOLD) {
+      return Instant.ofEpochMilli(timestamp).toString();
+    }
+    return Instant.ofEpochSecond(timestamp).toString();
+  }
+
+  public static class CursorField {
+
+    private static final Set<String> STRING_TYPES = Set.of(
+        "STRING", "CHAR", "NCHAR", "NVARCHAR", "VARCHAR", "LONGVARCHAR", "DATE",
+        "TIME", "TIMESTAMP"
+    );
+    private static final Set<String> NUMBER_TYPES = Set.of(
+        "NUMBER", "TINYINT", "SMALLINT", "INT", "INTEGER", "BIGINT", "FLOAT", "DOUBLE",
+        "REAL", "NUMERIC", "DECIMAL"
+    );
+
+    private final String name;
+    private final CursorType type;
+
+    public CursorField(String name, CursorType type) {
+      this.name = name;
+      this.type = type;
+    }
+
+    protected static CursorField fromStream(ConfiguredAirbyteStream stream) {
+      List<String> defaultCursorField = stream.getStream().getDefaultCursorField();
+      if (defaultCursorField == null || defaultCursorField.size() < 1){
+        return null;
+      }
+      String cursorField = defaultCursorField.get(0);
+      JsonNode typeNode = stream.getStream().getJsonSchema()
+          .get("properties").get(cursorField).get("type");
+      return new CursorField(cursorField, getType(typeNode));
+    }
+
+    private static CursorType getType(JsonNode typeNode) {
+      CursorType type = CursorType.UNRECOGNIZED;
+      if (typeNode.isArray()) {
```

Comment:
I've seen catalogs of some singer-based sources having fields types coded as [null, "number"]

Reply:
yup this is a good idea